framework/include/constraints/MortarSegmentHelper.h

```diff
@@ -0,0 +1,154 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+/**
+ * This class supports defining mortar segment mesh elements in 3D by projecting secondary and
+ * primary elements onto a linearized plane, computing the overlapping polygon formed by their
+ * projections, and triangulating the resulting nodes.
+ */
+
+class MortarSegmentHelper
+{
+public:
+  MortarSegmentHelper(const std::vector<Point> secondary_nodes,
+                      const Point & center,
+                      const Point & normal);
+
+  /**
+   * Computes the intersection between line segments defined by point pairs (p1,p2) and (q1,q2)
+   * Also computes s, the ratio of distance between (p1,p2) that the intersection falls,
+   * quantity s is useful in avoiding adding nearly degenerate nodes
+   */
+  Point getIntersection(
+      const Point & p1, const Point & p2, const Point & q1, const Point & q2, Real & s) const;
+
+  /**
+   * Check that a point is inside the secondary polygon (for verification only)
+   */
+  bool isInsideSecondary(const Point pt) const;
+
+  /**
+   * Checks whether polygons are disjoint for an easy out
+   */
+  bool isDisjoint(const std::vector<Point> & poly) const;
+
+  /**
+   * Clip secondary element (defined in instantiation) against given primary polygon
+   * result is a set of 2D nodes defining clipped polygon
+   */
+  std::vector<Point> clipPoly(const std::vector<Point> & primary_nodes) const;
+
+  /**
+   * Prints python plotting commands (to terminal) for visualizing polygons (for debugging only)
+   */
+  void plotPoly(const std::vector<Point> & poly) const;
+
+  /**
+   * Prints python plotting commands (to terminal) for visualizing triangulation (for debugging
+   * only)
+   */
+  void plotTriangulation(const std::vector<Point> & nodes,
+                         const std::vector<std::vector<unsigned int>> & elem_to_nodes) const;
+
```

Comment:
They aren't as necessary anymore, if you set _debug=true in AutomaticMortarGeneration it will output the mortar segment mesh as an exodus file. This can be useful for debugging the polygon clipping and stuff but I hope I've managed to make this pretty robust (it took a few iterations)

Reply:
Right. Only need to switch that flag in the constructor IIRC